larq_compute_engine/mlir/transforms/prepare_patterns_common.td

```diff
@@ -4,9 +4,24 @@ include "tensorflow/compiler/mlir/tensorflow/ir/tf_ops.td"
 include "larq_compute_engine/mlir/ir/lce_ops.td"
 include "larq_compute_engine/mlir/transforms/op_removal_patterns.td"
 
+class ConstantValue<string val> : AttrConstraint<CPred<"IsConstantValue($_self, " # val # ")">>;
 
-// This relies on implementation details of larq.math.sign. We should make
-// this more general in the future
+// Base quantiser patterns that match the `tf.where` implementation of `ste_sign`.
+def : Pat<(TF_SelectV2Op:$select_op
+              $cond,
+              (ConstantOp ConstantValue<"1.0f">),
+              (ConstantOp ConstantValue<"-1.0f">)),
+          (LQ_DequantizeOp (LQ_QuantizeOp $cond)),
+          [], (addBenefit 100)>;
+def : Pat<(TF_SelectV2Op:$select_op
+              $cond,
+              (ConstantOp ConstantValue<"-1.0f">),
+              (ConstantOp ConstantValue<"1.0f">)),
+          (LQ_DequantizeOp (LQ_QuantizeOp (TF_LogicalNotOp $cond))),
```

Comment:
Do you know if this logical not is folded with the `TFL_LessEqualOp` and `TFL_GreaterEqualOp` ? That is, will this also further trigger the optimization patterns?

Reply:
It seems to do so, yes :)

This is what I get when converting a model using the quantiser `lambda x: tf.where(x < 0.5, -1.0, 1.0)`:

![image](https://user-images.githubusercontent.com/7688302/123134386-a1125880-d448-11eb-875e-25fda9053025.png)